### PR TITLE
Fix | AE enclave retry logic not working for async queries

### PR DESCRIPTION
**Pull request by David-Engel**
This should address intermittent failures related to #1422 

Summary:

The retry logic when a cached enclave session expired (8 hours) doesn't look like it was implemented properly for async queries. There were several problems.

* The flow into the async retry path didn't work. The logic would simply bubble the RetryableEnclaveQueryExecutionException exception all the way up.
* It didn't make sense for the async retry to follow a different flow than the sync retry logic (which can handle an async flow just fine). 
* The multiple-thread handling code in EnclaveProviderBase didn't account for async tasks that might be running on different threads. The purpose of the logic is to block simultaneous threads requesting duplicate enclave sessions when one session could be used across them all if other threads were made to wait a bit. But when an async retry came through, it would introduce a 15 second wait since the retry task was on a different thread.
* During async retries, the async command count wasn't being incremented even though it had already been decremented during the original failure. Code comments indicated it was intentional at one point, but it isn't correct in the current flow.
* The retry mechanism for expired session caches wasn't being tested for expiry between commands nor for expiry during a narrow window between when we have the metadata to create the session and when we detect that a parameter will require generating an enclave package. It was this small window that was resulting in the `Enclave session is null during query execution.` errors. If the session cache entry expired at just the right time, the retry logic path would not even be hit.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlColumnEncryptionEnclaveProvider.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@ namespace Microsoft.Data.SqlClient
     internal abstract partial class SqlColumnEncryptionEnclaveProvider
     {
         /// <include file='../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlColumnEncryptionEnclaveProvider.xml' path='docs/members[@name="SqlColumnEncryptionEnclaveProvider"]/GetEnclaveSession/*'/>
-        internal abstract void GetEnclaveSession(EnclaveSessionParameters enclaveSessionParameters, bool generateCustomData, out SqlEnclaveSession sqlEnclaveSession, out long counter, out byte[] customData, out int customDataLength);
+        internal abstract void GetEnclaveSession(EnclaveSessionParameters enclaveSessionParameters, bool generateCustomData, out SqlEnclaveSession sqlEnclaveSession, out long counter, out byte[] customData, out int customDataLength, bool isRetry);
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlColumnEncryptionEnclaveProvider.cs]: Done

File: src/Microsoft.Data.SqlClient/tests/ManualTests/AlwaysEncrypted/ApiShould.cs [hunk 2]
```diff
@@ -2336,6 +2336,92 @@ public void TestRetryWhenAEParameterMetadataCacheIsStale(string connectionString
             cmd.ExecuteNonQuery();
         }
 
+        [ConditionalTheory(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringSetupForAE), nameof(DataTestUtility.EnclaveEnabled))]
+        [ClassData(typeof(AEConnectionStringProvider))]
+        public void TestRetryWhenAEEnclaveCacheIsStale(string connectionString)
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/AlwaysEncrypted/ApiShould.cs:2341]: This test calls both sync and async execute methods to cover both paths.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/AlwaysEncrypted/ApiShould.cs [hunk 3]
```diff
@@ -2336,6 +2336,92 @@ public void TestRetryWhenAEParameterMetadataCacheIsStale(string connectionString
...
+                    Assert.Equal(customerId, (int)reader[0]);
+                }
+                reader.Close();
+            }
+
+            CommandHelper.InvalidateEnclaveSession(cmd);
+
+            // Execute again to exercise the async session retry logic
+            Task readAsyncTask = ReadAsync(cmd, values, CommandBehavior.Default);
+            readAsyncTask.Wait();
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/AlwaysEncrypted/ApiShould.cs]: Done

**Review by Kaur-Parminder — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlCommand.cs [hunk 4]
```diff
@@ -2649,7 +2654,7 @@ private IAsyncResult BeginExecuteReaderInternal(CommandBehavior behavior, AsyncC
                                             // back into pool when we should not.
                 }
 
-                bool usedCache;
+                bool usedCache = false;
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlCommand.cs:2657]: Is this change necessary? the default value is false.

Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlCommand.cs:2657]: On the netcore side, if the assignment is left out, I get a compile error at the call to TriggerInternalEndAndRetryIfNecessary, where it is referenced. (`Error CS0165 Use of unassigned local variable 'usedCache'`) I don't get the error on the netfx side, but made the change in both places to keep the code in sync.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs [hunk 5]
```diff
@@ -2480,6 +2483,26 @@ long firstAttemptStart
             }, TaskScheduler.Default);
         }
 
+        private void InvalidateEnclaveSession()
+        {
+            if (ShouldUseEnclaveBasedWorkflow && this.enclavePackage != null)
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs:2488]: Save for a rainy day.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs [hunk 6]
```diff
@@ -2238,7 +2243,10 @@ private IAsyncResult BeginExecuteReaderInternal(CommandBehavior behavior, AsyncC
 
                     // For async, RunExecuteReader will never put the stateObj back into the pool, so do so now.
                     ReliablePutStateObject();
-                    throw;
+                    if (inRetry || e is not EnclaveDelegate.RetryableEnclaveQueryExecutionException)
+                    {
+                        throw;
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs:2248]: `usedCache` is a local variable. So definitely not that. I think the state is good as-is.

File: doc/snippets/Microsoft.Data.SqlClient/SqlColumnEncryptionEnclaveProvider.xml [hunk 7]
```diff
@@ -38,6 +38,7 @@ the enclave attestation protocol as well as the logic for creating and caching e
         <GetEnclaveSession>
             <param name="enclaveSessionParameters">The set of parameters required for enclave session.</param>
             <param name="generateCustomData"><see langword="true" /> to indicate that a set of extra data needs to be generated for attestation; otherwise, <see langword="false" />.</param>
+            <param name="isRetry">Indicates if this is a retry from a failed call.</param>
             <param name="sqlEnclaveSession">When this method returns, the requested enclave session or <see langword="null" /> if the provider doesn't implement session caching. This parameter is treated as uninitialized.</param>
             <param name="counter">A counter that the enclave provider is expected to increment each time SqlClient retrieves the session from the cache. The purpose of this field is to prevent replay attacks.</param>
             <param name="customData">A set of extra data needed for attestating the enclave.</param>
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply doc/snippets/Microsoft.Data.SqlClient/SqlColumnEncryptionEnclaveProvider.xml]: Fixed

**Review by JRahnama — APPROVED** [review]
